### PR TITLE
Rework ImmutableAttributes API

### PR DESCRIPTION
**Pull request by melix**
This PR changes the implementation of immutable attributes, by forcing the creation of such attributes through a factory. This allows us to cache the attributes and dramatically reduce the memory pressure around it.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultArtifactTransformTargets.java [hunk 1]
```diff
@@ -20,16 +20,22 @@
...
 import org.gradle.api.attributes.AttributeContainer;
 import org.gradle.api.internal.attributes.AttributeContainerInternal;
-import org.gradle.api.internal.attributes.DefaultAttributeContainer;
+import org.gradle.api.internal.attributes.DefaultMutableAttributeContainer;
+import org.gradle.api.internal.attributes.ImmutableAttributesFactory;
 
 import java.util.List;
 
 public class DefaultArtifactTransformTargets implements ArtifactTransformTargets {
+    private final ImmutableAttributesFactory cache;
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultArtifactTransformTargets.java]: Should we better use `attributesFactory` as field/parameter name consistently throughout the code base? Using `cache` here and there might confuse someone who is new to this. One might get the impressions these are two different concepts. (While it always is a factory which optimises memory usage by using caching.)

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildSessionScopeServices.java [hunk 2]
```diff
@@ -119,4 +121,8 @@ GeneratedGradleJarCache createGeneratedGradleJarCache(CacheRepository cacheRepos
         String gradleVersion = GradleVersion.current().getVersion();
         return new DefaultGeneratedGradleJarCache(cacheRepository, gradleVersion);
     }
+
+    ImmutableAttributesFactory createAttributesCache() {
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildSessionScopeServices.java]: Better also call this `createAttributesFactory()`?

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultArtifactTransformTargets.java (hunk 1 above)
Comment (melix) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultArtifactTransformTargets.java]: You're right, this is leftover from an initial implementation. I thought "factory" reflected the concept better than cache. I'll change those, thanks for spotting!

**Comment by melix** [comment]
I renamed the variables and merged into `master`. Thanks @jjohannes!

**Pull request merged**